File: LimSeoyeon/01_Stack/BOJ_2493_탑.cpp
#include <iostream>
#include <stack>

using namespace std;
int arr[500001];
int res[500001];
int main() {
	ios::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	int N;
	cin >> N;

	for (int i = 1; i <= N; i++) {
		cin >> arr[i]; // i번째 arr에는 i번째 탑의 높이가 저장됨.
	}  
	stack<pair<int,int>> st;

	st.push({ N,arr[N] }); //맨 마지막 탑은 신호를 받을 수 없으므로 바로 stack에 넣어놓음.

	for(int i=N-1; i>=1; i--) {
		while(!st.empty() && st.top().second<arr[i] ){// i번쨰 탑이 stack의 top에있는 탑의 신호를 받아줄 수 있는 상태.
			res[st.top().first] = i;
			st.pop();
		}
		st.push({ i,arr[i] });
	}
	for (int i = 1; i <= N; i++) {
		cout << res[i]<<' ';
	}

}

File: LimSeoyeon/01_Stack/BOJ_2493_탑.java
package com.ohgiraffers;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;
import java.util.StringTokenizer;

public class BOJ_2493 {
    public static class Pair{
      int first,second;
      Pair(int first, int second){
          this.first=first;
          this.second=second;
      }
    };

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer stk = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(stk.nextToken());

        int[] arr= new int[N+1];
        int[] res =new int[N+1];
        Stack<Pair> st = new Stack<>();

        stk = new StringTokenizer(br.readLine());
        for (int i=1; i<=N; i++){
            arr[i]= Integer.parseInt(stk.nextToken());
        }

        for(int i=N; i>=1; i--){
            while(!st.empty()&&st.peek().second<arr[i]){
                res[st.pop().first]=i;
            }
            st.push(new Pair(i,arr[i]));
        }


        StringBuilder sb = new StringBuilder();
        for (int i = 1; i <= N; i++) {
            sb.append(res[i]).append(" ");
        }
        System.out.println(sb.toString().trim());
    }
}
